src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -0,0 +1,182 @@
+import io.supertokens.pluginInterface.passwordless.PasswordlessDevice;
+import io.supertokens.pluginInterface.passwordless.UserInfo;
+import io.supertokens.pluginInterface.passwordless.exception.DuplicateCodeIdException;
+import io.supertokens.pluginInterface.passwordless.exception.DuplicateDeviceIdHashException;
+import io.supertokens.pluginInterface.passwordless.exception.DuplicateLinkCodeHashException;
+import io.supertokens.pluginInterface.passwordless.exception.DuplicatePhoneNumberException;
+import io.supertokens.pluginInterface.passwordless.exception.UnknownDeviceIdHash;
+import io.supertokens.pluginInterface.passwordless.sqlStorage.PasswordlessSQLStorage;
+import io.supertokens.storageLayer.StorageLayer;
+import io.supertokens.utils.Utils;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+import javax.annotation.Nullable;
+
+import org.apache.tomcat.util.codec.binary.Base64;
+
+public class Passwordless {
+    // We are storing the "alphabets" like this because we remove a few characters from the normal English alphabet.
+    // e.g.: remove easy to confuse chars (oO0, Il)
+    private static final String USER_INPUT_CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz";
+    private static final String USER_INPUT_CODE_NUM_CHARS = "123456789";
+
+    private static Character getRandomAlphaChar(SecureRandom generator) {
+        return USER_INPUT_CODE_ALPHABET.charAt(generator.nextInt(USER_INPUT_CODE_ALPHABET.length()));
+    }
+
+    private static Character getRandomNumChar(SecureRandom generator) {
+        return USER_INPUT_CODE_NUM_CHARS.charAt(generator.nextInt(USER_INPUT_CODE_NUM_CHARS.length()));
+    }
+
+    public static CreateCodeResponse createCode(Main main, String email, String phoneNumber, @Nullable String deviceId,
+            @Nullable String userInputCode) throws RestartFlowException, DuplicateLinkCodeHashException,
+            StorageQueryException, NoSuchAlgorithmException, InvalidKeyException {
+        PasswordlessSQLStorage passwordlessStorage = StorageLayer.getPasswordlessStorage(main);
+        if (deviceId == null) {
+            while (true) {
+                CreateCodeInfo info = CreateCodeInfo.generate(userInputCode);
+                try {
+                    passwordlessStorage.createDeviceWithCode(email, phoneNumber, info.code);
+
+                    return info.resp;
+                } catch (DuplicateLinkCodeHashException | DuplicateCodeIdException | DuplicateDeviceIdHashException e) {
+                    // These are retryable, so ignored here.
+                    // DuplicateLinkCodeHashException is also always retryable, because linkCodeHash depends on the
+                    // deviceId which is generated again during the retry
+                }
+            }
+        } else {
+            byte[] deviceIdBytes = Base64.decodeBase64(deviceId);
```

Comment:
Why not put this in the `generate` function too? This way, the `generate` function can take a `String deviceId` - cleaner.